File: apps/api/app/schemas/vendas_instagram.py
"""Schemas Pydantic do canal Instagram (Fase IG-1).

UPDATE omite FKs do pai (anti mass-assignment); segredos nunca são serializados
de volta (a API devolve ``*_set`` + máscara).
"""
from __future__ import annotations

import datetime
import uuid
from typing import Optional

from pydantic import BaseModel


# ── Config (credenciais IG, na vendas_disparo_config) ──────────────────────────
class InstagramConfigUpdate(BaseModel):
    instagram_user_id: Optional[str] = None
    instagram_username: Optional[str] = None
    instagram_verify_token: Optional[str] = None
    instagram_token: Optional[str] = None
    instagram_app_secret: Optional[str] = None
    clear_instagram_token: Optional[bool] = None
    clear_instagram_app_secret: Optional[bool] = None


class InstagramConfigPublic(BaseModel):
    instagram_user_id: Optional[str] = None
    instagram_username: Optional[str] = None
    instagram_verify_token: Optional[str] = None
    instagram_token_set: bool = False
    instagram_token_masked: Optional[str] = None
    instagram_app_secret_set: bool = False


# ── Gatilhos ───────────────────────────────────────────────────────────────────
class GatilhoCreate(BaseModel):
    palavra_chave: Optional[str] = None
    ativo: bool = True
    responder_publico: bool = True
    responder_dm: bool = False
    instrucao_ia: Optional[str] = None
    resposta_publica_fixa: Optional[str] = None


class GatilhoUpdate(BaseModel):
    palavra_chave: Optional[str] = None
    ativo: Optional[bool] = None
    responder_publico: Optional[bool] = None
    responder_dm: Optional[bool] = None
    instrucao_ia: Optional[str] = None
    resposta_publica_fixa: Optional[str] = None


class GatilhoPublic(BaseModel):
    id: uuid.UUID
    palavra_chave: Optional[str] = None
    ativo: bool
    responder_publico: bool
    responder_dm: bool
    instrucao_ia: Optional[str] = None
    resposta_publica_fixa: Optional[str] = None

    model_config = {"from_attributes": True}


# ── Comentários (leitura) ──────────────────────────────────────────────────────
class ComentarioPublic(BaseModel):
    id: uuid.UUID
    comment_id: str
    media_id: Optional[str] = None
    from_username: Optional[str] = None
    texto: Optional[str] = None
    lead_id: Optional[uuid.UUID] = None
    respondido_publico: bool
    respondido_dm: bool
    resposta_texto: Optional[str] = None
    erro: Optional[str] = None
    created_at: Optional[datetime.datetime] = None

    model_config = {"from_attributes": True}


# ── Posts (galeria — list_media) ───────────────────────────────────────────────
class PostPublic(BaseModel):
    id: str
    caption: Optional[str] = None
    media_type: Optional[str] = None
    media_url: Optional[str] = None
    permalink: Optional[str] = None
    timestamp: Optional[str] = None
    comments_count: Optional[int] = None


# ── Stats do canal ──────────────────────────────────────────────────────────────
class InstagramStats(BaseModel):
    comentarios: int
    respondidos: int
    leads: int
    erros: int


# ── Fase 2: comentários de um post + resposta manual ───────────────────────────
class ComentarioIG(BaseModel):
    id: Optional[str] = None
    text: Optional[str] = None
    username: Optional[str] = None
    timestamp: Optional[str] = None


class RespostaManual(BaseModel):
    texto: str
    publico: bool = True
    dm: bool = False
    from_username: Optional[str] = None


class RespostaManualResult(BaseModel):
    ok: bool
    respondido_publico: bool
    respondido_dm: bool

File: src/integrations/api/vendasInstagram.ts
/**
 * Client de Vendas — Automação Instagram (canal IG estilo ManyChat).
 * Bate com o backend FastAPI (`/vendas/instagram/*`, snake_case). Usa o client
 * interno `@/integrations/api/client` (cookie httpOnly + refresh em 401).
 * Segredos (token, app_secret) nunca voltam em claro — só `*_set` + `*_masked`.
 */
import { api } from "@/integrations/api/client";

export interface InstagramConfigPublic {
  instagram_user_id: string | null;
  instagram_username: string | null;
  instagram_verify_token: string | null;
  instagram_token_set: boolean;
  instagram_token_masked: string | null;
  instagram_app_secret_set: boolean;
}

export interface InstagramConfigUpdate {
  instagram_user_id?: string | null;
  instagram_username?: string | null;
  instagram_verify_token?: string | null;
  instagram_token?: string | null;
  instagram_app_secret?: string | null;
  clear_instagram_token?: boolean | null;
  clear_instagram_app_secret?: boolean | null;
}

export interface Gatilho {
  id: string;
  palavra_chave: string | null;
  ativo: boolean;
  responder_publico: boolean;
  responder_dm: boolean;
  instrucao_ia: string | null;
  resposta_publica_fixa: string | null;
}

export interface GatilhoInput {
  palavra_chave?: string | null;
  ativo?: boolean;
  responder_publico?: boolean;
  responder_dm?: boolean;
  instrucao_ia?: string | null;
  resposta_publica_fixa?: string | null;
}

export interface Comentario {
  id: string;
  comment_id: string;
  media_id: string | null;
  from_username: string | null;
  texto: string | null;
  lead_id: string | null;
  respondido_publico: boolean;
  respondido_dm: boolean;
  resposta_texto: string | null;
  erro: string | null;
  created_at: string | null;
}

export interface Post {
  id: string;
  caption: string | null;
  media_type: string | null;
  media_url: string | null;
  permalink: string | null;
  timestamp: string | null;
  comments_count: number | null;
}

export interface ComentarioIG {
  id: string | null;
  text: string | null;
  username: string | null;
  timestamp: string | null;
}

export interface RespostaManualPayload {
  texto: string;
  publico: boolean;
  dm: boolean;
  from_username?: string | null;
}

export interface InstagramStats {
  comentarios: number;
  respondidos: number;
  leads: number;
  erros: number;
}

export const vendasInstagramApi = {
  getConfig: () => api.get<InstagramConfigPublic>("/vendas/instagram/config"),
  saveConfig: (data: InstagramConfigUpdate) =>
    api.put<InstagramConfigPublic>("/vendas/instagram/config", data),

  getStats: () => api.get<InstagramStats>("/vendas/instagram/stats"),

  listGatilhos: () => api.get<Gatilho[]>("/vendas/instagram/gatilhos"),
  createGatilho: (data: GatilhoInput) =>
    api.post<Gatilho>("/vendas/instagram/gatilhos", data),
  updateGatilho: (id: string, data: GatilhoInput) =>
    api.put<Gatilho>(`/vendas/instagram/gatilhos/${id}`, data),
  deleteGatilho: (id: string) =>
    api.del<void>(`/vendas/instagram/gatilhos/${id}`),

  listComentarios: (limit = 50) =>
    api.get<Comentario[]>(`/vendas/instagram/comentarios?limit=${limit}`),

  listPosts: () => api.get<Post[]>("/vendas/instagram/posts"),

  listComentariosPost: (mediaId: string) =>
    api.get<ComentarioIG[]>(`/vendas/instagram/posts/${mediaId}/comentarios`),
  responderComentario: (commentId: string, data: RespostaManualPayload) =>
    api.post<{ ok: boolean; respondido_publico: boolean; respondido_dm: boolean }>(
      `/vendas/instagram/comentarios/${commentId}/responder`, data),
};
